core/src/main/scala/org/bykn/bosatsu/Normalization.scala

```diff
@@ -0,0 +1,303 @@
+      normalBranches = branches.map { case (p, e) => (p, e.tag._2.ne)}
+      neTag = NormalExpressionTag(ne=NormalExpression.Match(arg.tag._2.ne, normalBranches), children=Set())
+    }
+    yield Match(arg=arg,
+      branches=branches,
+      tag=(m.tag, neTag)
+      )
+  }
+
+  def normalizeBranch(b: (Pattern[(PackageName, Constructor), Type], TypedExpr[Declaration]), env: Env, p: Package.Inferred): State[
+  Map[(PackageName, Identifier), ResultingRef],
+  (Pattern[(PackageName, Constructor), Type], TypedExpr[(Declaration, NormalExpressionTag)])] = {
+    val (pattern, expr) = b
+    val names = pattern.names.collect { case b: Identifier.Bindable => b}.map(Some(_))
+    val lambdaVars = names ++ env._2
+    val nextEnv = (env._1 ++ lambdaVars.zipWithIndex
+      .collect { case (Some(n), i) => (n, i) }
+      .toMap
+      .mapValues(idx => NormalExpressionTag(NormalExpression.LambdaVar(idx), Set[NormalExpression]())),
+      lambdaVars)
+    for {
+      innerExpr <- normalizeExpr(expr, nextEnv, p)
+      normalExpr = names.foldLeft(innerExpr.tag._2.ne) { case (expr, _) => NormalExpression.Lambda(expr) }
+      finalExpression = innerExpr.updatedTag((innerExpr.tag._1, innerExpr.tag._2.copy(ne=normalExpr)))
+    } yield (pattern, finalExpression)
+
+  }
+
+  def normalizePackageLet(pkgName: PackageName, inferredExpr: (Identifier.Bindable, RecursionKind, TypedExpr[Declaration]), pack: Package.Inferred): State[
+  Map[(PackageName, Identifier), ResultingRef], (Identifier.Bindable, RecursionKind, TypedExpr[(Declaration, Normalization.NormalExpressionTag)])] = {
+    println(inferredExpr)
+
+    println(s"let ${inferredExpr._1} = ${inferredExpr._3}")
+    for {
+      expr <- normalizeExpr(inferredExpr._3, (Map(), Nil), pack)
+      _ <- State.modify { cache: Map[(PackageName, Identifier), ResultingRef] => cache + ((pkgName, inferredExpr._1) -> Right(expr))}
+    }
+      yield (inferredExpr._1, inferredExpr._2, expr)
+  }
+
+  def normalizeProgram(pkgName: PackageName, pack: Package.Inferred): State[
+  Map[(PackageName, Identifier), ResultingRef],
+  Program[TypeEnv[Variance], TypedExpr[(Declaration, Normalization.NormalExpressionTag)], Statement]] = {
+    for { 
+      lets <- pack.program.lets.map(normalizePackageLet(pkgName, _, pack)).sequence
+    } yield pack.program.copy(
+      lets  = lets
+    )
+  }
+
+  def normalizePackage(pkgName: PackageName, pack: Package.Inferred): State[
+  Map[(PackageName, Identifier), ResultingRef],
+  Package.Normalized] = for {
+    program <- normalizeProgram(pkgName, pack)
+  } yield pack.copy(program = program)
+
+  def getTag(ref: ResultingRef) = ref match {
+    case Right(te) => te.tag
+    case Left((_, t)) => t
+  }
+
+  private type Ref[T] =
+    Either[(Identifier, T), TypedExpr[T]]
+
+  private type SourceRef = Ref[Declaration]
+  private type ResultingRef = Ref[(Declaration,  NormalExpressionTag)]
+  private type Env = (Map[Identifier, NormalExpressionTag], List[Option[Identifier]])
```

Comment:
can we comment the meaning of `List[Option[Identifier]]`? I don't know what the None means, nor why there is a List there.

Reply:
I'm also 90% sure I can get rid of the `Option`. This was just because when `Identifier`s were just `String` and when you cared about them being Bindable there were `Option[String]`. Now we just have `Bindable` as being explicit.

I'll add a comment but I'll also explain here. This is here for the de bruijn indexing substitutions. The context is the `Env` is only here to produce a `NormalExpression` given a `TypedExpr`. When an identifier is encountered it looks it up in `_1` of the `Env`. Sometimes that will map to `NormalExpression.LambdaVar(2)` for example, eg when you're referencing the identifier of the argument of a lambda.

But, when you define a new lambda inside (which you do all the time because of alpha regularization) all the `LambdaVar` indices have to be incremented. This is where the `List` comes in. It is just a `List` of identifiers of `Lambda` args. By appending to it each time we start a new `Lambda` we have exactly the de bruijn index as the list index.